Enterprise GitHub

rcaileff opened the pull request:
GitHub Enterprise Support #99
This implementation is working for me.  Let me know if there are pieces that need updating that I didn't realize were relevant. :)

czechboy0 commented:
Thanks @rcaileff for this work! It's looking great overall, I just need to clear up one thing:

As far as I understand, when you're running an Enterprise server, each server will have different keys/secrets, so if at all possible, I'd like to avoid users using classic OAuth flow (because we'd need to ask users to create an developer application on their GitHub Enterprise server, copy its id/secret into Buildasaur and only then do the actual authentication). I'd much more prefer users use the simple [Personal Token](https://github.com/settings/tokens) flow. Since users will have to do something manually anyway, creating and copying one Personal Token to Buildasaur is easier than copying key/secret and then doing authentication. Plus, we won't have to add UI for it.

Is there any reason to not use Personal Tokens? If you're fine with it, BuildasaurKeys won't need any extra keys inside. Because the server's URL should be parsed out of the project's remote URL (and somehow we should ping the server to recognize it as a valid GitHub Enterprise server as validation). I'll add comments to the places that I believe you don't need to change if you just use a Personal Token.

Does this make sense? If you're not sure about how the authentication works exactly, GitHub has [great docs](https://developer.github.com/v3/oauth/) on it (search for Personal Token).

czechboy0 commented on BuildaGitServer/GitServerPublic.swift:
```diff
@@ -14,12 +14,14 @@ import Result
 
 public enum GitService: String {
     case GitHub = "github"
+    case EnterpriseGitHub = "enterprisegithub"
```
With this addition, I think we should refactor GitService to the following

```
public enum GitService {
    case GitHub
    case EnterpriseGitHub(url: String)
    case BitBucket
}
```

Moving to an associated value only for the Enterprise case will make things much easier.

czechboy0 commented on BuildaGitServer/GitServerPublic.swift:
```diff
@@ -34,6 +37,7 @@ public enum GitService: String {
     public func hostname() -> String {
         switch self {
         case .GitHub: return "github.com"
+        case .EnterpriseGitHub: return BuildasaurKeys().enterpriseGitHubHostname()
```
If you do the refactoring of the GitService enum, then here you can just return the associated value as the hostname. BuildasaurKeys cannot store this information.

czechboy0 commented on BuildaGitServer/GitServerPublic.swift:
```diff
@@ -48,6 +53,7 @@ public enum GitService: String {
     public func accessTokenUrl() -> String {
         switch self {
         case .GitHub: return "https://github.com/login/oauth/access_token"
+        case .EnterpriseGitHub: return "https://\(hostname())/login/oauth/access_token"
```
I think we'll disallow the classic OAuth flow for Enterprise for now, so feel free to add an assert here (when GitService is EnterpriseGitHub, it should never be asked for accessTokenUrl)

czechboy0 commented on BuildaGitServer/GitServerPublic.swift:
```diff
@@ -55,6 +61,7 @@ public enum GitService: String {
     public func serviceKey() -> String {
         switch self {
         case .GitHub: return BuildasaurKeys().gitHubAPIClientId()
+        case .EnterpriseGitHub: return BuildasaurKeys().enterpriseGitHubAPIClientId()
```
Same here

czechboy0 commented on BuildaGitServer/GitServerPublic.swift:
```diff
@@ -62,6 +69,7 @@ public enum GitService: String {
     public func serviceSecret() -> String {
         switch self {
         case .GitHub: return BuildasaurKeys().gitHubAPIClientSecret()
+        case .EnterpriseGitHub: return BuildasaurKeys().enterpriseGitHubAPIClientSecret()
```
And same here

czechboy0 commented on BuildaGitServerTests/GitHubServerTests.swift:
```diff
@@ -14,11 +14,13 @@ import BuildaUtils
 class GitHubSourceTests: XCTestCase {
 
     var github: GitHubServer!
+    var enterpriseGithub: GitHubServer!
```
It might be worth separating EnterpriseGithub tests into a separate file and only test the things not already tested in GitHub tests.

czechboy0 commented on Buildasaur/ProjectViewController.swift:
```diff
@@ -167,7 +167,7 @@ class ProjectViewController: ConfigEditViewController {
 
         switch service {
-        case .GitHub:
+        case .GitHub, .EnterpriseGitHub:
```
In light of disallowing OAuth flow for Enterprise, you'll need to add an option here of _only_ allowing the token field to be shown (and hide the OAuth login button).

czechboy0 commented on Buildasaur/ServiceAuthentication.swift:
```diff
@@ -66,6 +66,8 @@ class ServiceAuthenticator {
         switch service {
         case .GitHub:
             return self.getGitHubParameters()
+        case .EnterpriseGitHub:
```
Remove this and let it assert if Enterprise gets asked for params, OAuth won't be allowed for now

czechboy0 commented on Buildasaur/ServiceAuthentication.swift:
```diff
@@ -73,6 +75,25 @@ class ServiceAuthenticator {
         }
     }
     
+    private func getEnterpriseGitHubParameters() -> ([ParamKey: String], SecretFromResponseParams) {
```
You can remove this whole function

czechboy0 commented on Podfile:
```diff
@@ -4,6 +4,10 @@ plugin 'cocoapods-keys', {
+    "EnterpriseGitHubHostname", # e.g. git.mycompany.com
+    "EnterpriseGitHubAPIPath", # e.g. git.mycompany.com/api/v1"
+    "EnterpriseGitHubAPIClientId",
+    "EnterpriseGitHubAPIClientSecret",
```
Here you don't need to add these, because 
- hostname will be parsed from the repo's remote URL
- API path will need to be discovered dynamically
- OAuth won't be supported, so the ClientId and Secret are not necessary.

czechboy0 commented on fastlane/Fastfile:
```diff
@@ -43,8 +43,12 @@ end
+      'EnterpriseGitHubHostname',
+      'EnterpriseGitHubAPIPath',
+      'EnterpriseGitHubAPIClientId',
+      'EnterpriseGitHubAPIClientSecret',
```
Same here, since we're not changing them in BuildasaurKeys, we don't need to mock them here

czechboy0 commented:
Sorry for bombarding you with those comments, I just wanted to make it easier to go and change the things to disallow OAuth for Enterprise, both in code and in UI.

As far as I see it, we'll need to add
- parsing of server host+API path from a git repo's remote URL
- pinging the new server to verify it's a valid Enterprise server

Let me know if you have any questions :) Thanks for working on this. I don't have/use Enterprise, so I'm glad more users will be able to take advantage of Buildasaur.

czechboy0 commented:
Also, please merge from master, I fixed the Keys issue there.

rcaileff commented:
Awesome, thanks.

On Sun, Mar 6, 2016 at 12:09 PM, Honza Dvorsky <notifications@github.com>
wrote:

> Also, please merge from master, I fixed the Keys issue there.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/czechboy0/Buildasaur/pull/248#issuecomment-192951639>.
>

rcaileff commented:
I rebased from master and now I have a large number of build errors,
including "Use of unresolved identifier 'BuildasaurKeys' ".  :(

On Mon, Mar 7, 2016 at 8:35 AM, Caileff, Rachel <rcaileff@enova.com> wrote:

> Awesome, thanks.
>
> On Sun, Mar 6, 2016 at 12:09 PM, Honza Dvorsky <notifications@github.com>
> wrote:
>
>> Also, please merge from master, I fixed the Keys issue there.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/czechboy0/Buildasaur/pull/248#issuecomment-192951639>
>> .
>>
>
>

czechboy0 commented:
It's probably still cached, try to delete
- `~/.cocoapods/keys`
- `Pods`
and open Keychain Access and search for any keys starting with **Cocoapods-Keys**, delete those as well
- delete derived data

Then run pod install again. Should work, works for me both on CI and locally.

czechboy0 commented:
I'll close this since you opened up #251.
